### ISSUE TITLE
OfMany relationship with non-numeric Primary Key breaks the query

### ISSUE THREAD

**Issue by danilopolani** [laravel/framework#38904]
- Laravel Version: 8.61.0
- PHP Version: 8.0.9
- Database Driver & Version: Pgsql (12.x)

### Description:
When using a relationship to a table using a non-numeric ID/Primary Key (e.g. a UUID), the `ofMany` relationship breaks the query because it always appends the primary key of the model to the query ([reference](https://github.com/laravel/framework/blob/3d1ead403e05ee0b9aa93a6ff720704970aec9c8/src/Illuminate/Database/Eloquent/Relations/Concerns/CanBeOneOfMany.php#L85-L87)), even if you set a custom column name.

A fix could be to append the PK only if it's an integer using the `getKeyType()` method of the model. I think it should be safe and backward compatible, since if you're using a string as primary key it won't work anyway and it won't affect the common use-case (integers).

### Steps To Reproduce:
1. Create a table `users`, the columns don't matter
1. Create a table `orders` with a UUID as primary ID and a Foreign Key `user_id` pointing to `users`
1. In the `User` model, create a relationship to get the latest order:

```php
public function latestRun(): HasOne
{
    return $this->hasOne(Order::class)->latestOfMany('created_at');
}
```

Now loading the relationship will throw an error similar to this:

```
ERROR: function max(uuid) does not exist LINE 1: select * from "orders" inner join (select MAX("id") a... ^
HINT: No function matches the given name and argument types. You might need to add explicit type casts.

SQL: select * from "orders" inner join (select MAX("id") as "id", "orders"."user_id" from "orders" inner join (select MAX("created_at") as "created_at"...
```

**Comment by danilopolani** [comment]
I've tried right now with MySQL and it works fine, I guess it does something under-the-hood to convert the string to int(?), so the problem is restricted to PgSQL only.

**Comment by derekmd** [comment]
Previously reported: https://github.com/laravel/framework/issues/37854

Laravel docs have a specific callout for Postgres not supporting `MAX()` on UUID columns:

* https://laravel.com/docs/8.x/eloquent-relationships#has-one-of-many
* https://github.com/laravel/docs/pull/7170

**Comment by danilopolani** [comment]
> Previously reported: #37854
> 
> Laravel docs have a specific callout for Postgres not supporting `MAX()` on UUID columns:
> 
> * https://laravel.com/docs/8.x/eloquent-relationships#has-one-of-many
> * [[8.x] Clarify UUID's and one-of-many with PostgreSQL docs#7170](https://github.com/laravel/docs/pull/7170)

Uh, I think I missed that, thank you.

**Comment by lazosweb** [comment]
Since PgSQL does not support MAX(UUID) we can have a secondary column in the DB that is incremental and we can use that one for the ofMany queries in PgSQL or you find this approach "hacky"?

Unfortunately, we cannot do this because the framework is forcing the primary key to be in the query when we could provide a different column for max...
```php
$keyName = $this->query->getModel()->getKeyName();

$columns = is_string($columns = $column) ? [
    $column => $aggregate,
    $keyName => $aggregate,
] : $column;

if (! array_key_exists($keyName, $columns)) {
    $columns[$keyName] = 'MAX';
}
```

**Comment by danilopolani** [comment]
> Since PgSQL does not support MAX(UUID) we can have a secondary column in the DB that is incremental and we can use that one for the ofMany queries in PgSQL or you find this approach "hacky"?
> 
> The issue we cannot do this currently is that the framework is forcing the primary key to be in the query when we could provide a different column for max...
> 
> $keyName = $this->query->getModel()->getKeyName();
> 
> $columns = is_string($columns = $column) ? [
>     $column => $aggregate,
>     $keyName => $aggregate,
> ] : $column;
> 
> if (! array_key_exists($keyName, $columns)) {
>     $columns[$keyName] = 'MAX';
> }

Yeah don't know why it forces us to use always use the PK. Probably it's worth to add another param to the relationship just to handle these scenarios and avoiding using always the PK but just the custom column

**Comment by lazosweb** [comment]
> > Since PgSQL does not support MAX(UUID) we can have a secondary column in the DB that is incremental and we can use that one for the ofMany queries in PgSQL or you find this approach "hacky"?
> > The issue we cannot do this currently is that the framework is forcing the primary key to be in the query when we could provide a different column for max...
> > $keyName = $this->query->getModel()->getKeyName();
> > $columns = is_string($columns = $column) ? [
> > $column => $aggregate,
> > $keyName => $aggregate,
> > ] : $column;
> > if (! array_key_exists($keyName, $columns)) {
> > $columns[$keyName] = 'MAX';
> > }
> 
> Yeah don't know why it forces us to use always use the PK. Probably it's worth to add another param to the relationship just to handle these scenarios and avoiding using always the PK but just the custom column

I found the why on this post https://github.com/laravel/framework/pull/37362#pullrequestreview-666248254 
That code is there to prevent duplications and I get it, but still I am expecting a way to override it and provide a different unique column...
@cbl any chance to change that part that forces the primary key so that developers can provide a different unique key? It will not resolve the UUID issue on PgSQL but we can bypass it this way won't we?

**Comment by aimeos** [comment]
As MAX/MIN isn't supported by PostgreSQL for UUID columns, using a SELECT with ORDER BY and LIMIT 1 would be preferable instead (https://github.com/laravel/framework/discussions/58497)
